File: src/mhe.cpp
#include "turtlebot_MHE/mhe.h"
#include <ceres/ceres.h>
#include <fstream>

template<typename T>
T wrap(T angle)
{
    static double pi{3.14159};
    angle -= 2*pi * floor((angle+pi) * 0.5/pi);
    return angle;
}

struct PoseResidual
{
public:
    PoseResidual(const Eigen::Vector3d &x, const Eigen::Matrix3d &omega): x_{x}
    {
        xi_ = omega.llt().matrixL();
    }

    template<typename T>
    bool operator()(const T* const x, T* residual) const
    {
        Eigen::Map<const Eigen::Matrix<T,3,1>> pose(x);
        Eigen::Map<Eigen::Matrix<T,1,3>> res(residual);
        Eigen::Matrix<T,3,1> temp{x_ - pose};
        temp(mhe::THETA) = wrap(temp(mhe::THETA));
        res = temp.transpose() * xi_;
        return true;
    }
protected:
    Eigen::Vector3d x_;
    Eigen::Matrix3d xi_;

};

struct MeasurementResidual
{
public:
    MeasurementResidual(const Eigen::Vector2d &z, const Eigen::Vector2d &lm, const Eigen::Matrix2d &R_inv): z_{z}, lm_{lm}
    {
        xi_ = R_inv.llt().matrixL();
    }

    template<typename T>
    bool operator()(const T* const x, T* residual) const
    {
        Eigen::Map<const Eigen::Matrix<T,3,1>> pose(x);
        Eigen::Matrix<T,2,1> diff; // = lm_ - pose.segment<2>(mhe::X);
        diff << lm_(mhe::X) - pose(mhe::X), lm_(mhe::Y) - pose(mhe::Y);
        Eigen::Map<Eigen::Matrix<T,2,1>> res(residual);
        T range = diff.norm();
        T phi = wrap(atan2(diff(mhe::Y), diff(mhe::X)) - pose(mhe::THETA));
        Eigen::Matrix<T,2,1> z_hat{range, phi};
        Eigen::Matrix<T, 2, 1> temp{z_ - z_hat};
        temp(mhe::PHI) = wrap(temp(mhe::PHI));
        
        res = temp.transpose() * xi_;
        return true;
    }

protected:
    Eigen::Vector2d z_, lm_;
    Eigen::Matrix2d xi_;
};

struct OdomResidual
{
public:
    OdomResidual(const Eigen::Vector3d &x, const Eigen::Matrix3d &omega): x_{x}
    {
        xi_ = omega.llt().matrixL();
    }

    template<typename T>
    bool operator()(const T* const x1, const T* const x2, T* residual) const
    {
        Eigen::Map<const Eigen::Matrix<T,3,1>> pose1(x1), pose2(x2);
        Eigen::Map<Eigen::Matrix<T,1,3>> res(residual);
        Eigen::Matrix<T,3,1> diff{pose2 - pose1};
        diff(mhe::THETA) = wrap(diff(mhe::THETA));
        Eigen::Matrix<T,3,1> temp{x_ - diff};
        temp(mhe::THETA) = wrap(temp(mhe::THETA));
        res = temp.transpose() * xi_;
        return true;
    }
protected:
    Eigen::Vector3d x_;
    Eigen::Matrix3d xi_;

};

typedef ceres::AutoDiffCostFunction<PoseResidual,3,3> PoseCostFunction;
typedef ceres::AutoDiffCostFunction<MeasurementResidual, 2, 3> MeasurementCostFunction;
typedef ceres::AutoDiffCostFunction<OdomResidual,3,3,3> OdomCostFunction;

namespace mhe
{

MHE::MHE()
{
    // set default parameters
    Omega_ = Eigen::Vector3d{1, 1, 0.5}.asDiagonal();
    R_inv_ = Eigen::Vector2d{1/0.35, 1/0.07}.asDiagonal();
    lms_ = Meas::Zero();
    mu_ = Pose{2.5, -1.7, 0.175}; //These values from lab. Put these in the params file
    pose_hist_.push_back(mu_);
}

MHE::~MHE()
{
    writeFile();
}

void MHE::setParams(const Eigen::Vector3d &omega, double sig_r, double sig_phi)
{
    Omega_ = omega.asDiagonal();
    R_inv_ = Eigen::Vector2d{1/(sig_r*sig_r),1/(sig_phi*sig_phi)}.asDiagonal();
}

Pose MHE::propagateState(const Pose &state, const Input &u, double dt)
{
    double st{sin(state(THETA))};
    double ct{cos(state(THETA))};
    Pose out;
    out(X) = u(V) * ct * dt;
    out(Y) = u(V) * st * dt;
    out(THETA) = u(W) * dt;
    odom_hist_.push_back(out);
    out = state + out;
    out(THETA) = wrap(out(THETA));
    return out;
}

void MHE::update(const Pose &mu, const Meas &z, const Zidx& idx, const Input &u, double dt)
{
//    Pose mu_bar{propagateState(mu, u, dt)};
    mu_ = propagateState(mu_, u, dt);

    // pose_hist_.push_back(mu);
    pose_hist_.push_back(mu_);
    z_hist_.push_back(z);
    z_ind_ = idx;

    optimize();
}

void MHE::optimize()
{
    ceres::Problem problem;

    //set up position residuals
    int i = std::max(0, int(pose_hist_.size() - TIME_HORIZON));
    for(i; i < pose_hist_.size(); ++i)
    {
        PoseCostFunction *cost_function{new PoseCostFunction(new PoseResidual(pose_hist_[i], Omega_))};
        problem.AddResidualBlock(cost_function, NULL, pose_hist_[i].data());
    }

//    //set up odometry residuals
//    i = std::max(0, int(odom_hist_.size() - TIME_HORIZON));
//    Eigen::Matrix3d Om = Eigen::Vector3d{1e1, 1e1, 5e-1}.asDiagonal();
//    for(i; i < odom_hist_.size(); ++i)
//    {
//        OdomCostFunction * cost_function{new OdomCostFunction(new OdomResidual(odom_hist_[i], Omega_))};
//        problem.AddResidualBlock(cost_function, NULL, pose_hist_[i].data(), pose_hist_[i+1].data());
//    }

    //set up measurement residuals
    i = std::max(0, int(z_hist_.size() - TIME_HORIZON));
    int counter = 0;
    for(i; i < z_hist_.size(); ++i) // May need nested for loops for this one. One for the index and then one for each measurement at that index
    {
        for(int j{0}; j < NUM_LANDMARKS; ++j)
        {
//            if(z_ind_(TIME_HORIZON - counter - 1,j))
            if(z_ind_(counter,j))
            {
//                if(z_hist_[i].col(j)(0) == 0.0 || z_hist_[i].col(j)(1) == 0.0)
//                    int debug =  1;
                //Need the true landmark locations to be stored somewhere
                MeasurementCostFunction *cost_function{new MeasurementCostFunction(new MeasurementResidual(z_hist_[i].col(j), lms_.col(j), R_inv_))};
                problem.AddResidualBlock(cost_function, NULL, pose_hist_[i+1].data());
            }
        }
        ++counter;
    }

    //setup options and solve
    ceres::Solver::Options options;
    options.minimizer_progress_to_stdout = false;
    options.max_num_iterations = 50;
    options.gradient_tolerance = 1e-8;
    options.function_tolerance = 1e-8;
    options.linear_solver_type = ceres::SPARSE_NORMAL_CHOLESKY;
    ceres::Solver::Summary summary;
    ceres::Solve(options, &problem, &summary);
}

void MHE::initializeLandmark(int index, const Eigen::Vector2d &lm)
{
    lms_.col(index) = lm;
}

void MHE::writeFile()
{
    std::ofstream file;
    file.open("/tmp/MHE_landmarks.txt");
    file << lms_.transpose();
    file.close();

    file.open("/tmp/MHE_outputs.txt");
    for (Pose pose : pose_hist_)
        file << pose.transpose() << std::endl;
    file.close();
}
} // namespace mhe

File: python/mhe.py
import numpy as np
from copy import deepcopy
from scipy.optimize import minimize
from extractdata import landmarks
from copy import deepcopy

def unwrap(phi):
    phi -= 2 * np.pi * np.floor((phi + np.pi) * 0.5/np.pi)
    return phi

class MHE:
    def __init__(self):
        self.pose_hist = []    
        self.z_hist = []       
        self.z_ind_hist = []

        self.mu = np.array([2.5,-1.7, 0.175])

        self.pose_hist.append(deepcopy(self.mu)) #change this so that the position is inisialized by the constructor
        self.Omega = np.diag([1e3, 1e3, 0.5e3])
        self.Omega2 = np.diag([1e3, 1e3, 0.5e3])
        self.R_inv = np.diag([1/(.35**2), 1/(.07**2)])
        self.N = 5  #Size of the window to optimize over

    def setParams(self, omega, sig_r, sig_phi):
        self.Omega = omega
        self.R_inv = np.diag([1/(sig_r**2), 1/(sig_phi**2)])

    def propagateState(self, state, v, w, dt):
        theta = state[2]
        st = np.sin(theta)
        ct = np.cos(theta)

        A = np.array([v * ct,
                    v * st,
                    w])
        temp = state + A * dt
        temp[2] = unwrap(temp[2])
        return temp

    def update(self, mu, z, z_ind, v, w, dt):
        mu_bar = self.propagateState(self.mu, v, w, dt)

        self.pose_hist.append(mu_bar)
        self.z_hist.append(z)
        self.z_ind_hist.append(z_ind)

        if len(self.pose_hist) >= self.N + 1: 
            mu_temp = self.optimize(self.pose_hist[-self.N:], self.z_hist[-self.N:], self.z_ind_hist[-self.N:])
            mu_temp = mu_temp.reshape((3, int(mu_temp.size/3)), order='F')
            mu_temp[2] = unwrap(mu_temp[2])
            self.mu = mu_temp[:,-1]
            for i in range(self.N):
                self.pose_hist[-(self.N - i)] = mu_temp[:,i]
        else:
            mu_temp = self.optimize(self.pose_hist, self.z_hist, self.z_ind_hist) # z_hist doesn't have the same length as pose_hist here b/c original position doesn't have
            mu_temp = mu_temp.reshape((3, int(mu_temp.size/3)), order='F')
            mu_temp[2] = unwrap(mu_temp[2])
            self.mu = mu_temp[:,-1]
            for i in range(int(mu_temp.size/3)):
                self.pose_hist[i] = mu_temp[:,i]

    def optimize(self, mu, z, z_ind):
        mu = np.array(mu).T.flatten(order='F')
        x0 = deepcopy(mu)
        sigma = np.array(sigma)

        x_hat_opt = minimize(self.objective_fun, mu, method='SLSQP', jac=False, args=(x0, z, z_ind, landmarks), options={'ftol':1e-5, 'disp':False})

        return x_hat_opt.x

    def objective_fun(self, mu, x0, z, z_ind, lms):
        e_x = 0
        e_z = 0

        dx = (x0 - mu).reshape((-1, 3, 1), order='F')
        dx[:,2] = unwrap(dx[:,2])
        e_x = np.sum(dx.transpose(0,2,1) @ self.Omega @ dx) #Hand tuned values

        # temp = mu.reshape((-1,3,1), order='F')
        # dx2 = np.diff(temp, axis=0)
        # temp2 = x0.reshape((-1,3,1),order='F')
        # dx0 = np.diff(temp2, axis=0)
        # diff = dx0 = dx2
        # diff[:,2] = unwrap(diff[:,2])
        # e_x2 = np.sum((dx0 - dx2).transpose(0,2,1) @ self.Omega2 @ (dx0 - dx2)) #Error between successive poses

        for i in range(len(z)):
            if z_ind[i].size > 0:
                z_hat = self.h(mu[3*i:3*i+3], lms, z_ind[i])
                dz = z[i] - z_hat
                dz[1] = unwrap(dz[1])
                e_z += np.sum(np.diagonal(dz.T @ self.R_inv @ dz))

        return e_x + e_z

    def h(self, mu, lms, z_ind): #Need to check if this works
        mu_temp = mu.reshape((3, int(mu.size/3)), order='F')
        lm = lms[:,z_ind.squeeze()].reshape((2,-1))
        ds = lm - mu_temp[0:2]
        r = np.sqrt(np.sum(ds*ds, axis=0))
        theta = np.arctan2(ds[1], ds[0]) - mu_temp[2]
        theta = unwrap(theta)
        z_hat = np.vstack((r,theta))

        return z_hat
